ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -233,17 +395,23 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
     } else
         return (ret);
 
+    /* Create the file system. */
+    if ((fs = (S3_FILE_SYSTEM *)calloc(1, sizeof(S3_FILE_SYSTEM))) == NULL)
+        return (errno);
+    fs->storage = s3;
+
     /* Store a copy of the home directory in the file system. */
     homeDir = session->connection->get_home(session->connection);
+    fs->homeDir = homeDir;
+
+    /* Fetch the native WT file system. */
+    if ((ret = s3->wtApi->file_system_get(s3->wtApi, session, &wtFileSystem)) != 0)
+        return (ret);
```

Comment:
I reshuffled the code such that we fetch all the configuration and only after succeeding on doing that allocate a filesystem. In this change, if you fail at line 409, you would have allocated memory for the file system at line 399, without freeing on error return, hence leaking memory.

I will suggest allocate memory for the filesystem at the very last, after having obtained anything needed to initialize the file system with.

Reply:
Agreed; will reshuffle so that all assignments to the file system are made after it has been allocated and everything needed has been fetched.